### PR TITLE
FederatedDataset & DataLoader improvement

### PR DESCRIPTION
**Pull request by LaRiffle**
Make these objects more similar to their dual Pytorch implementation, and more easy to use

### CODE REVIEW

File: syft/frameworks/torch/federated/__init__.py [hunk 1]
```diff
@@ -1,2 +1,2 @@
-from .dataset import FederatedDataset
+from .dataset import FederatedDataset, BaseDataset
```
Comment (robert-wagner) [inline syft/frameworks/torch/federated/__init__.py]: Separate lines pls

File: syft/frameworks/torch/federated/dataset.py [hunk 2]
```diff
@@ -1,59 +1,109 @@
...
-            inputs (dict): dict of dataset tensors or pointers to dataset tensors
-            targets (dict): dict of target/label tensors or pointers to target/label tensors
-            batch_size (int): size of each batch
-            num_iterators (int): number of iterators to use for parallel computations
-        """
+import torch
+from torch.utils.data import Dataset
+
+
+class BaseDataset:
```
Comment (robert-wagner) [inline syft/frameworks/torch/federated/dataset.py]: Docstring?

File: syft/frameworks/torch/federated/dataset.py [hunk 3]
```diff
@@ -1,59 +1,109 @@
...
-            self.worker2num_rows[worker] = worker_inputs.shape[0]
-            self.worker2inputs[worker] = worker_inputs
-            self.worker2targets[worker] = worker_targets
 
-            self.num_rows += worker_inputs.shape[0]
+Dataset.federate = dataset_federate
+BaseDataset.federate = dataset_federate
+
+
+class FederatedDataset:
```
Comment (robert-wagner) [inline syft/frameworks/torch/federated/dataset.py]: Docstring

File: test/torch/federated/test_dataset.py [hunk 4]
```diff
@@ -1,10 +1,96 @@
...
+    dataset.get()
+    try:
+        assert dataset.data.location.id == 0
+        assert dataset.targets.location.id == 0
+        assert False
+    except AttributeError:
+        pass
 
 
 def test_federated_dataset(workers):
```
Comment (robert-wagner) [inline test/torch/federated/test_dataset.py]: Federated class?

Comment (LaRiffle) [reply test/torch/federated/test_dataset.py]: Sorry?

Comment (robert-wagner) [reply test/torch/federated/test_dataset.py]: Should have been more clear. All these tests important base dataset. Should they he using federated dataset?

Comment (LaRiffle) [reply test/torch/federated/test_dataset.py]: I'm not sure I get all.
Basically this tests the BaseDataset class, which is supposed to be a class containing two tensors (.data and .targets) and which you can manipulate (.send(bob), .get() ) just like a tensor

Comment (LaRiffle) [reply test/torch/federated/test_dataset.py]: This is what to give to feed a FederatedDataset

**Pull request merged**